docs/src/nondiff_points.md

```diff
@@ -0,0 +1,141 @@
+`3` is the arguably the nicest, but it is also the most expensive to compute.
+In general all are acceptable.
+
+
+### Derivative zero almost everywhere
+
+```@example nondiff
+plot(ceil)
+```
+
+Here it is most useful to say the derivative is zero everywhere.
+The limits are zero from both sides.
+
+The other option for `x->ceil(x)` would be relax the problem into `x->x`, and thus  say it is 1 everywhere
+But that it too weird, if the use wanted a relaxation of the problem then they would provide one.
+We can not be imposing that relaxation on to `ceil` for everyone is not reasonable.
+
+### Not defined on one-side
+```@example nondiff
+plot(x->exp(2log(x)))
+plot!(; xlims=(-10,10), ylims=(-10,10)) #hide
+```
+
+We do not have to worry about what to return for the side where it is not defined.
+As we will never be asked for the derivative at e.g. `x=-2.5` since the primal function errors.
+But we do need to worry about at the boundary -- if that boundary point doesn't error.
+
+Since we will never be asked about the left-hand side (as the primal errors), we can use just the right-hand side derivative.
+In this case giving 0.0.
+`
+Also nice in this case is that it agrees with the symbolic simplification of `x->exp(2log(x))` into `x->x^2`.
+
+
+### Derivative nonfinite and same on both sides
+
+```@example nondiff
+plot(cbrt)
+```
+
+Here we have no real choice but to say the derivative at `0` is `Inf`.
+We could consider as an alternative saying some large but finite value.
+However, if too large it will just overflow rapidly anyway; and if too small it will not dominate over finite terms.
+It is not possible to find a given value that is always large enough.
+Our alternatives  woud be to consider the dederivative at `nextfloat(0.0)` or `prevfloat(0.0)`.
+But this is more or less the same as choosing some large value -- in this case an extremely large value that will rapidly overflow.
+
+
+### Derivative on-finite and different on both sides
+
+```@example nondiff
+plot(x-> sign(x) * cbrt(x))
+```
+
+In this example, the primal is defined and finite, so we would like a derivative to defined.
+We are back in the case of a local minimal like we were for `abs`.
+We can make most of the same arguments as we made there to justify saying the derivative is zero.
+
+## Conclusion
+
+From the case studies a few general rules can be seen for how to choose a value that is _useful_.
+These rough rules are:
+ - Say the derivative is 0 at local optima
+ - If the derivative from one side is defined and the other isn't, say it is the derivative taken from defined side.
+ - If the derivative from one side is finite and the other isn't, say it is the derivative taken from finite side.
+ - When derivative from each side is not equal, strongly consider reporting the average
+
+Our goal as always, is to get a pragmatically useful result for everyone, which must by necessity also avoid a pathological result for anyone.
```

Comment:
Maybe worth mentioning that we can't always get the result that's best for literally everyone, but we sometimes just have to do our best.